crates/libafl/src/corpus/dynamic.rs

```diff
@@ -0,0 +1,141 @@
+//! Dynamic corpus that allows users to switch corpus types at runtime.
+
+use core::{cell::RefCell, marker::PhantomData};
+
+use libafl_bolts::Error;
+use serde::{Deserialize, Serialize};
+
+use crate::corpus::{Corpus, CorpusId, Testcase};
+
+/// An dynamic corpus type accepting two types of corpus at runtime. This helps rustc better
+/// reason about the bounds compared to dyn objects.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum DynamicCorpus<C1, C2, I> {
+    /// Corpus1 implementation
+    Corpus1(C1, PhantomData<I>),
+    /// Corpus2 implementation
+    Corpus2(C2, PhantomData<I>),
+}
+
+/// A helper macro to avoid too many duplicates.
+macro_rules! select_corpus {
+    ($sf: tt, $it: tt, $( $args: tt), *) => {
+        match $sf {
+            Self::Corpus1(c1, _) => c1.$it( $($args),* ),
+            Self::Corpus2(c2, _) => c2.$it( $($args),* ),
+        }
+    };
+}
+
+impl<C1, C2, I> DynamicCorpus<C1, C2, I>
+where
+    C1: Corpus<I>,
+    C2: Corpus<I>,
+{
+    /// Create a `DynamicCorpus` with Corpus1 variant.
+    pub fn corpus1(c: C1) -> Self {
+        Self::Corpus1(c, PhantomData)
+    }
+
+    /// Create a `DynamicCorpus` with Corpus2 variant.
+    pub fn corpus2(c: C2) -> Self {
+        Self::Corpus2(c, PhantomData)
+    }
+}
+
+impl<C1, C2, I> Corpus<I> for DynamicCorpus<C1, C2, I>
+where
+    C1: Corpus<I>,
+    C2: Corpus<I>,
+{
+    fn peek_free_id(&self) -> CorpusId {
+        select_corpus!(self, peek_free_id,)
```

Comment:
I don't really think it's worth it to do a macro here. It's only three lines of code anyway. Macro is less legible IMHO, and has compile time impact

Reply:
Make sense. Will fix.